sdcm/nemesis.py

```diff
@@ -5008,6 +5011,115 @@ def disrupt_disable_binary_gossip_execute_major_compaction(self):
+        try:
+            with simulate_node_unavailability(self.target_node):
+                # target node stopped by Contextmanger. Wait while its status will be updated
+                wait_for(node_operations.is_node_seen_as_down, timeout=600, throw_exc=True,
+                         down_node=self.target_node)
+                self.log.debug("Remove node %s : hostid: %s with stopped scylla from cluster",
+                               self.target_node.name, target_host_id)
+                working_node.run_nodetool(f"removenode {target_host_id}", retry=3)
+                self.log.debug("Wait while all node update topology status")
+                wait_for(node_operations.is_node_removed_from_cluster, timeout=300, throw_exc=True,
+                         node=self.target_node)
+                # Run read barrier on working node to validate that it update raft topology state
+                self.cluster.call_read_barrier(working_node)
+
+            # Context manager at exit  start scylla on target node.
+            # But node already removed from cluster. So any operations from it
+            # should be banned. If query executed succesfull, raise an error
+            assert self.target_node.is_port_used(
+                port=9042, service_name="scylla-service"), f"Port 9042 is closed on node {self.target_node.name}"
+
+            with self.cluster.cql_connection_exclusive(node=self.target_node) as session:
+                try:
+                    stmt = SimpleStatement("INSERT INTO banned_keyspace.table1 (id, name) VALUES (1, 'name1');",
+                                           consistency_level=ConsistencyLevel.ALL)
+                    session.execute(stmt)
+                    raise BannedQueryExecutedSuccessful(
+                        "Query from banned node was executed succesful with Consistency.ALL")
+                except BannedQueryExecutedSuccessful:
+                    self.log.error("Banned query passed to cluster from banned node")
+                    raise
+                except (NoHostAvailable, OperationTimedOut, Unavailable) as exc:
+                    self.log.debug("Query failed with error: %s as expected", exc)
+
+        except Exception as exc:  # pylint: disable=broad-except
+            self.log.error("Method failed with error %s", exc)
+            raise
+        finally:
+            # because target node removed from cluster, terminate it
+            # and bootstrap new one
+            LOGGER.debug("Terminating node %s", self.target_node.name)
+            self.cluster.terminate_node(self.target_node)
+            # check node was removed previously, if not remove it from cluster
+            removed_node_status = self.cluster.get_node_status_dictionary(
+                ip_address=self.target_node.ip_address, verification_node=working_node)
+            if removed_node_status is not None:
+                working_node.run_nodetool(f"removenode {target_host_id}", retry=3)
```

Comment:
Why do we expect 2 possible states after running `nodetool removenode` on the line 5059?
And why we are ok with both?

Reply:
Once we trigger removenode, the node marked as non-voter and banned from cluster. The banned node and rest nodes in the cluster could see each other with different state. So if for any reason any step is failed or passed (failed to remove node, not blocking ports, or anything else) or node was removed successfully.   in finally block, i want to restore the cluster health. for that i need to remove target node totally from cluster and add new one. So if any reason target node stay in cluster in main nemesis flow, it will be removed from cluster in finally block and doesn't matter in which state it was.